source/common/ssl/context_manager_impl.cc

```diff
@@ -20,23 +21,93 @@ void ContextManagerImpl::releaseContext(Context* context) {
+      map_exact_[listener_name][EMPTY_STRING] = nullptr;
+    }
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
+        if (map_wildcard_[listener_name][name] == context) {
+          map_wildcard_[listener_name][name] = nullptr;
+        }
+      } else {
+        if (map_exact_[listener_name][name] == context) {
+          map_exact_[listener_name][name] = nullptr;
+        }
+      }
+    }
+  }
+
+  // context may not be found, in the case that a subclass of Context throws
+  // in it's constructor.  In that case the context did not get added, but
+  // the destructor of Context will run and call releaseContext().
+  contexts_.remove(context);
+}
+
 ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
                                                             ClientContextConfig& config) {
-
   ClientContextPtr context(new ClientContextImpl(*this, scope, config));
   std::unique_lock<std::mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
   return context;
 }
 
-ServerContextPtr ContextManagerImpl::createSslServerContext(Stats::Scope& scope,
-                                                            ServerContextConfig& config) {
-  ServerContextPtr context(new ServerContextImpl(*this, scope, config, runtime_));
+ServerContextPtr
+ContextManagerImpl::createSslServerContext(const std::string& listener_name,
+                                           const std::vector<std::string>& server_names,
+                                           Stats::Scope& scope, ServerContextConfig& config) {
+  ServerContextPtr context(
+      new ServerContextImpl(*this, listener_name, server_names, scope, config, runtime_));
   std::unique_lock<std::mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
+
+  // Save mappings.
+  if (server_names.empty()) {
+    map_exact_[listener_name][EMPTY_STRING] = context.get();
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
+        map_wildcard_[listener_name][name] = context.get();
+      } else {
+        map_exact_[listener_name][name] = context.get();
+      }
+    }
+  }
+
   return context;
 }
 
+ServerContext* ContextManagerImpl::findSslServerContext(const std::string& listener_name,
+                                                        const std::string& server_name) {
+  std::unique_lock<std::mutex> lock(contexts_lock_);
+  if (map_exact_[listener_name][server_name] != nullptr) {
+    return map_exact_[listener_name][server_name];
+  }
+
+  // Try to construct and match wildcard domain.
+  if (server_name.size() >= 5) {
+    size_t pos = server_name.find('.');
+    if (pos > 0) {
+      size_t rpos = server_name.rfind('.');
+      if (rpos > pos + 1 && rpos != server_name.size() - 1) {
+        std::string wildcard = '*' + server_name.substr(pos);
```

Comment:
nit: any reason to not use the wildcard matching algorithm that we have for virtual hosts?

Reply:
+1 to reusing the matching code in config_impl. Can we factor that out into a utility?

Reply:
Which code do you mean, exactly? `ContextImpl::dNSNameMatch`?

Reply:
source/common/router/config_impl.h, RouteMatcher does similar work.  See "domains" in https://www.envoyproxy.io/envoy/configuration/http_conn_man/route_config/vhost.html#config-http-conn-man-route-table-vhost.  Without a strong argument against, I think it makes sense for the matching semantics to be the same between this code and that.

Reply:
I'd +1 as well for the reason that we might change the implementation to be smarter one day, e.g. trie-based as per @tschroed 's earlier implementation. Probably not needed right now, but just as future proofing.